### ISSUE TITLE
6.0.0.beta1

### ISSUE THREAD

**Issue by bizantl** [CanCanCommunity/cancancan#556]
### Steps to reproduce
After upgrading to 6.0.0.beta1, the app fails to respond properly. No matter what I do, i get an error:
NotImplemented, 'This model adapter does not support fetching records from the database.'

### System configuration
**Rails version**:
6.0.0.beta1
**Ruby version**:
2.6.0
**CanCanCan version**
2.3.0

**Comment by francesco-loreti** [comment]
I can confirm, same error

**Comment by coorasse** [comment]
I'll schedule a fix for the version 3.0.0 since Rails 6 is still in Beta. I plan to release it by the end of February.

**Comment by coorasse** [comment]
Issue can now be seen in #560

**Comment by francesco-loreti** [comment]
Thank you for your faster job

**Comment by phoet** [comment]
when running rails 6 beta1 some code-paths that use `accessible_by` are failing with `PostgreSQL with DISTINCT ORDER raises PG::InvalidColumnReference`

i'm not sure if this is a regression of https://github.com/rails/rails/issues/31783 or something in cancancan. (or maybe im holding it wrong)

**Comment by coorasse** [comment]
@phoet can you share the steps to reproduce the issue?

**Comment by phoet** [comment]
@coorasse the issue is with this commit: https://github.com/CanCanCommunity/cancancan/blame/feature/3.0.0/lib/cancan/model_adapters/active_record_5_adapter.rb 

before
```
      # As of rails 4, `includes()` no longer causes active record to
      # look inside the where clause to decide to outer join tables
      # you're using in the where. Instead, `references()` is required
      # in addition to `includes()` to force the outer join.
      def build_relation(*where_conditions)
        relation = @model_class.where(*where_conditions)
        relation = relation.includes(joins).references(joins) if joins.present?
        relation
      end
```

after
```
      def build_relation(*where_conditions)
        relation = @model_class.where(*where_conditions)
        relation = relation.left_joins(joins).distinct if joins.present?
        relation
      end
```

in combination with an `order` call to a join table, this now fails as the join-table is not within the select statement.

query before
```
SELECT "companies"."name" AS t1_r22,  [...] FROM "staff_members" LEFT OUTER JOIN "companies" [...] ORDER BY companies.name, last_name
```

query after
```
SELECT DISTINCT "staff_members".* FROM "staff_members" [...] ORDER BY companies.name, last_name
```

**Comment by coorasse** [comment]
I see. cancancan won't take care of calling includes and joins for you anymore from version 3.0.0. 
I guess calling them explicitly could solve your issue. Could you also share the relevant parts of ruby code with the `order` and `accessible_by` calls and your abilities definition? This will help me, thanks!

**Comment by phoet** [comment]
@coorasse i think this wont work. using `distinct` does not allow including the joined table into the select.

so currently i have code that is like this:

```
broker.staff_members.accessible_by(current_ability).includes(:company).order('companies.name')
```

**Comment by phoet** [comment]
was playing around with this minimal rails example to find out whats going on https://gist.github.com/phoet/6d42e95028869ef55cfc6bd3a0973489

**Comment by coorasse** [comment]
The problem of ordering by an association attribute is not new: you want to retrieve staff_members but order them by the company name. since we join the companies table, we need to use `distinct` to avoid duplicate results, but doing so, we don't allow ordering by a different attribute. Maybe adding a `select('"staff_members".*, "companies.name"')` could fix it but I need to write a proper regression test and find a solution. Thanks!

**Comment by phoet** [comment]
yeah, as a workaound `.select('"staff_members".*, "companies"."name"')` does the trick (made a small adjustment).

**Comment by coorasse** [comment]
Because of how is going to work cancancan 3.0 the behavior of `order` on associations will be marked as a breaking change. The cases are two:
* one-to-one relationships: you need to explicitly write the `select` statement as explained [here](https://github.com/CanCanCommunity/cancancan/issues/556#issuecomment-460219397)
* one-to-many relationships: you should not `order` by an association attribute.

CanCanCan could recognise the first case and adapt the query but this is not going to happen: I think is out of scope for an authorization gem. Is more important to know how the gem works and how are the queries generated, and how to deal with this situation.

**Comment by coorasse** [comment]
This issue is currently solved in 3.0.0. I close it. Thanks everyone! 🙇 ❤️
